Add PyPI release workflow (#9297)
dask/distributed: docs/release-procedure.md

Distributed follows a similar procedure for releasing as the core Dask project.

See https://github.com/dask/dask/blob/main/docs/release-procedure.md for instructions.

Pushing a release tag to `dask/distributed` triggers
`.github/workflows/release-publish.yml`, which builds and smoke-tests the wheel
and source distribution, verifies that they depend on the matching Dask release,
publishes them to PyPI with Trusted Publishing, and publishes the GitHub Release.

The workflow can be rehearsed from GitHub Actions without publishing to PyPI by
running `Release Publisher` manually. Provide the Distributed version to
rehearse and a Dask version that is already available on PyPI. Manual runs
build distributions, check versions, run `twine check`, upload and download
workflow artifacts, wait for the requested Dask version on PyPI, run wheel and
source-distribution smoke tests, and run a dry-run publish job. They do not
enter the protected PyPI environment, upload to PyPI, or publish GitHub
Releases.

> For coordinated Dask and Distributed releases, the Dask and Distributed tags may

Reviewer: ```suggestion
To save time, the Dask and Distributed tags may
```
The wording may suggest that you can have one without the other, while you don't

be pushed together. The Distributed workflow waits until the matching
`dask==YYYY.M.X` wheel and source distribution are available on PyPI before
smoke-testing and publishing. The Distributed smoke tests install dependencies
from PyPI and assert that the installed Dask version matches the release, so the
matching Dask release must be resolvable first.

> GitHub Actions pauses at the `pypi` environment for manual approval. Open the

Reviewer: I can't find this pause in the `publish_pypi` job. Is it an implicit default in `pypa/gh-action-pypi-publish@release`? If so I'd rather make it explicit.
